### PR TITLE
Moved Lease Client to specialized package

### PR DESCRIPTION
**Pull request by gapra-msft**
Resolves https://github.com/Azure/azure-sdk-for-java/issues/6317
Also adds in a check to make sure endpoints passed into builders are dfs endpoints

### CODE REVIEW

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/implementation/util/DataLakeImplUtils.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.storage.file.datalake.implementation.util;
+
+public class DataLakeImplUtils {
+    public static String endpointToDesiredEndpoint(String endpoint, String desiredEndpoint, String currentEndpoint) {
+        // Add the . on either side to prevent overwriting an endpoint if a user provides a
+        String desiredRegex = "." + desiredEndpoint + ".";
```
Comment (jaschrep-msft) [inline sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/implementation/util/DataLakeImplUtils.java]: please don't call this a regex

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/specialized/DataLakeLeaseClientBuilder.java [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+    /**
+     * Initializes a {@link SpecializedBlobClientBuilder}
+     * @param dfsEndpoint The endpoint for the {@link SpecializedBlobClientBuilder}
+     * @param pipeline The {@link HttpPipeline} for the {@link SpecializedBlobClientBuilder}
+     * @return the {@link SpecializedBlobClientBuilder}
+     */
+    private SpecializedBlobClientBuilder getSpecializedBlobClientBuilder(String dfsEndpoint, HttpPipeline pipeline) {
+        String blobEndpoint = BlobUrlParts.parse(
+            DataLakeImplUtils.endpointToDesiredEndpoint(dfsEndpoint, "blob", "dfs")).toUrl().toString();
```
Comment (jaschrep-msft) [inline sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/specialized/DataLakeLeaseClientBuilder.java]: make this simple like the others.

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/specialized/DataLakeLeaseClientBuilder.java [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+    /**
+     * Initializes a {@link BlobContainerClientBuilder}
+     * @param dfsEndpoint The endpoint for the {@link BlobContainerClientBuilder}
+     * @param pipeline The {@link HttpPipeline} for the {@link BlobContainerClientBuilder}
+     * @return the {@link BlobContainerClientBuilder}
+     */
+    private BlobContainerClientBuilder getBlobContainerClientBuilder(String dfsEndpoint, HttpPipeline pipeline) {
+        String blobEndpoint = BlobUrlParts.parse(
+            DataLakeImplUtils.endpointToDesiredEndpoint(dfsEndpoint, "blob", "dfs")).toUrl().toString();
```
Comment (jaschrep-msft) [inline sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/specialized/DataLakeLeaseClientBuilder.java]: same as above

File: sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/APISpec.groovy [hunk 4]
```diff
@@ -88,7 +90,7 @@ class APISpec extends Specification {
 
     static final String garbageLeaseID = UUID.randomUUID().toString()
 
-    public static final String defaultEndpointTemplate = "https://%s.dfs.core.windows.net/"
+    public static final String defaultEndpointTemplate = "http://%s.dfs.core.windows.net/"
```
Comment (jaschrep-msft) [inline sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/APISpec.groovy]: revert

Comment (rickle-msft) [inline sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/APISpec.groovy]: Revert

File: sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/APISpec.groovy [hunk 5]
```diff
@@ -161,15 +163,15 @@ class APISpec extends Specification {
...
-        String testMode = Configuration.getGlobalConfiguration().get(AZURE_TEST_MODE)
-
-        if (testMode != null) {
-            try {
-                return TestMode.valueOf(testMode.toUpperCase(Locale.US))
-            } catch (IllegalArgumentException ignore) {
-                return TestMode.PLAYBACK
-            }
-        }
+//        String testMode = Configuration.getGlobalConfiguration().get(AZURE_TEST_MODE)
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/APISpec.groovy]: Revert

File: sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/APISpec.groovy [hunk 6]
```diff
@@ -161,15 +163,15 @@ class APISpec extends Specification {
...
+//
+//        if (testMode != null) {
+//            try {
+//                return TestMode.valueOf(testMode.toUpperCase(Locale.US))
+//            } catch (IllegalArgumentException ignore) {
+//                return TestMode.PLAYBACK
+//            }
+//        }
```
Comment (jaschrep-msft) [inline sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/APISpec.groovy]: revert

Comment (gapra-msft) [reply sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/APISpec.groovy]: reverted

File: sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/APISpec.groovy (hunk 5 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/APISpec.groovy]: reverted

File: sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/APISpec.groovy (hunk 4 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/APISpec.groovy]: reverted

Comment (gapra-msft) [reply sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/APISpec.groovy]: reverted

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/specialized/DataLakeLeaseClientBuilder.java (hunk 2 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/specialized/DataLakeLeaseClientBuilder.java]: done

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/specialized/DataLakeLeaseClientBuilder.java (hunk 3 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/specialized/DataLakeLeaseClientBuilder.java]: done

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/implementation/util/DataLakeImplUtils.java (hunk 1 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/implementation/util/DataLakeImplUtils.java]: changed to stringToMatch

**Pull request merged**